File: app/src/main/java/kr/hs/emirim/sookhee/redonorpets/adapter/ShelterAdapter.java
package kr.hs.emirim.sookhee.redonorpets.adapter;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.HashMap;

import kr.hs.emirim.sookhee.redonorpets.R;
import kr.hs.emirim.sookhee.redonorpets.model.ShelterData;
import kr.hs.emirim.sookhee.redonorpets.ui.shelter.ShelterProfileActivity;

public class ShelterAdapter extends RecyclerView.Adapter<ShelterAdapter.CustomViewHolder> {

    private Context mCtx;
    private HashMap<String, ShelterData> mData;
    private ArrayList<String> shelterPosition = new ArrayList<>();

    public ShelterAdapter(Context mCtx) {
        this.mCtx = mCtx;
        mData = new HashMap<>();
    }

    @NonNull
    @Override
    public CustomViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(mCtx).inflate(R.layout.item_shelter, parent, false);

        return new CustomViewHolder(v);
    }

    @Override
    public void onBindViewHolder(@NonNull CustomViewHolder holder, final int position) {
        ShelterData shelter = (ShelterData) mData.values().toArray()[position];

        String img = shelter.getProfileImg();
        holder.shelterPositionInAdpater = shelter.getShelterPosition();

        holder.tvShelterName.setText(shelter.getName());
        holder.tvShelterPhone.setText(shelter.getPhone());
        Picasso.get().load(img).into(holder.ivShelterProfile);
    }

    @Override
    public int getItemCount() {
        return mData.size();
    }

    class CustomViewHolder extends RecyclerView.ViewHolder{

        TextView tvShelterName;
        TextView tvShelterPhone;
        ImageView ivShelterProfile;
        View pView;
        String shelterPositionInAdpater;


        public CustomViewHolder(View itemView) {
            super(itemView);

            pView = itemView;
            tvShelterName = itemView.findViewById(R.id.shelterNameTextView);
            tvShelterPhone = itemView.findViewById(R.id.shelterPhoneTextView);
            ivShelterProfile = itemView.findViewById(R.id.shelterProfileImageView);

            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Intent intent;
                    intent = new Intent(v.getContext(), ShelterProfileActivity.class);
                    intent.putExtra("shelterPosition", shelterPositionInAdpater);
                    v.getContext().startActivity(intent);
                }
            });
        }

    }

    public void addDataAndUpdate(String key, ShelterData p){
        mData.put(key, p);
        notifyDataSetChanged();
    }

    public void deleteDataAndUpdate(String key){
        mData.remove(key);
        notifyDataSetChanged();
    }
}

File: app/src/main/java/kr/hs/emirim/sookhee/redonorpets/ui/shelter/ShelterAdapter.kt
package kr.hs.emirim.sookhee.redonorpets.ui.shelter

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import kr.hs.emirim.sookhee.redonorpets.common.setImageWithUrl
import kr.hs.emirim.sookhee.redonorpets.databinding.ItemShelterBinding
import kr.hs.emirim.sookhee.redonorpets.databinding.ItemShelterGridBinding
import kr.hs.emirim.sookhee.redonorpets.domain.entity.Shelter

/**
 *  ShelterAdapter.kt
 *
 *  Created by Minji Jeong on 2022/01/12
 *  Copyright © 2022 DonorPets2.0. All rights reserved.
 */

class ShelterAdapter :
    RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private var items: List<Shelter> = emptyList()

    var shelterViewType = SHELTER_VIEW_TYPE.GRID
    var onItemClick: ((selectedItem: Shelter) -> Unit)? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val inflater = LayoutInflater.from(parent.context)

        return when (shelterViewType) {
            SHELTER_VIEW_TYPE.GRID -> ShelterGridViewHolder(
                ItemShelterGridBinding.inflate(inflater, parent, false),
                onItemClick
            )
            SHELTER_VIEW_TYPE.LIST -> ShelterListViewHolder(
                ItemShelterBinding.inflate(inflater, parent, false),
                onItemClick
            )
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (shelterViewType) {
            SHELTER_VIEW_TYPE.GRID -> (holder as ShelterGridViewHolder).bind(items[position])
            SHELTER_VIEW_TYPE.LIST -> (holder as ShelterListViewHolder).bind(items[position])
        }
    }

    override fun getItemCount(): Int {
        return items.size
    }

    @SuppressLint("NotifyDataSetChanged")
    fun setItem(list: List<Shelter>) {
        items = list
        notifyDataSetChanged()
    }

    class ShelterGridViewHolder(
        private val binding: ItemShelterGridBinding,
        onItemClick: ((selectedItem: Shelter) -> Unit)?,
    ) :
        RecyclerView.ViewHolder(binding.root) {

        init {
            binding.onItemClick = onItemClick
        }

        @SuppressLint("SetTextI18n")
        fun bind(
            item: Shelter,
        ) {
            binding.item = item

            binding.tvShelterName.text = item.name
            binding.tvStoryCount.text = "${item.storyCount}개의 스토리"
            binding.ivShelterProfile.setImageWithUrl(item.profileImage)
        }
    }

    class ShelterListViewHolder(
        private val binding: ItemShelterBinding,
        onItemClick: ((selectedItem: Shelter) -> Unit)?,
    ) :
        RecyclerView.ViewHolder(binding.root) {

        init {
            binding.onItemClick = onItemClick
        }

        @SuppressLint("SetTextI18n")
        fun bind(
            item: Shelter,
        ) {
            binding.item = item

            binding.ivShelterProfile.setImageWithUrl(item.profileImage)
        }
    }

    companion object {
        enum class SHELTER_VIEW_TYPE { GRID, LIST }
    }
}
